### PR TITLE
Include NLA header on mcast groups and ops.

### PR DESCRIPTION
**Pull request by dylanswiggett**
The emit logic for McastGroups doesn't include the required NLA header on nested NLAs. Without this change, netlink-packet-generic emits incorrect buffers if an mcast group is included, and cannot parse the bytes it writes.

### CODE REVIEW

**Review by cathay4t — CHANGES_REQUESTED** [review]
This is not elegant and prone to errors when there is padding.

How about:

* `GenlCtrlAttrs::McastGroups(Vec<McastGroup>)`
* `struct McastGroup { nlas: Vec<McastGrpAttrs>}`
* impl Nla for `McastGroup`

With this, we can have `nlas.as_slice().buffer_len()` and `nlas.as_slice().emit(buffer)` saving us from worry about headers.

Meanwhile, please fix `CTRL_ATTR_OPS` also.

**Comment by dylanswiggett** [comment]
Sure, pushed an updated version and fixed ops too.

File: src/ctrl/nlas/mcast.rs [hunk 1]
```diff
@@ -4,13 +4,42 @@ use crate::constants::*;
...
 use std::mem::size_of_val;
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct McastGroup {
+    pub nlas: Vec<McastGrpAttrs>,
+}
+
+impl Nla for McastGroup {
+    fn value_len(&self) -> usize {
+        self.nlas.iter().map(|nla| nla.buffer_len()).sum()
```
Comment (cathay4t) [inline src/ctrl/nlas/mcast.rs:21]: Please use `self.nlas.as_slice().buffer_len()`

File: src/ctrl/nlas/mcast.rs [hunk 2]
```diff
@@ -4,13 +4,42 @@ use crate::constants::*;
...
+    }
+
+    fn kind(&self) -> u16 {
+        NLA_F_NESTED
```
Comment (cathay4t) [inline src/ctrl/nlas/mcast.rs:25]: The kind() is not `NLA_F_NESTED`. It is `index + 1`. The kernel code is:

```c
nla_grps = nla_nest_start_noflag(skb, CTRL_ATTR_MCAST_GROUPS);
if (nla_grps == NULL)
	goto nla_put_failure;

for (i = 0; i < family->n_mcgrps; i++) {
	struct nlattr *nest;
	const struct genl_multicast_group *grp;

	grp = &family->mcgrps[i];

	nest = nla_nest_start_noflag(skb, i + 1); // <<<<<< HERE
	if (nest == NULL)
		goto nla_put_failure;
```

Please check `BondIpAddrNlaList` in `netlink-packet-route` as an example.

**Review by cathay4t — CHANGES_REQUESTED** [review]
The `kind()` implementation is wrong.

Meanwhile, please also include unit test case to test the `emit` and `parse` using the binary captured by `strace` or `nlmon`.

File: src/ctrl/nlas/ops.rs [hunk 3]
```diff
@@ -4,13 +4,42 @@ use crate::constants::*;
...
+    pub nlas: Vec<OpAttrs>,
+}
+
+impl Nla for Op {
+    fn value_len(&self) -> usize {
+        self.nlas.iter().map(|nla| nla.buffer_len()).sum()
+    }
+
+    fn kind(&self) -> u16 {
+        NLA_F_NESTED
```
Comment (cathay4t) [inline src/ctrl/nlas/ops.rs:25]: This should be index of Op. This is kernel code:

```c
nla_ops = nla_nest_start_noflag(skb, CTRL_ATTR_OPS);
if (nla_ops == NULL)
	goto nla_put_failure;

while (genl_op_iter_next(&i)) {
	struct nlattr *nest;
	u32 op_flags;

	op_flags = i.flags;
	if (i.doit.policy || i.dumpit.policy)
		op_flags |= GENL_CMD_CAP_HASPOL;

	nest = nla_nest_start_noflag(skb, genl_op_iter_idx(&i));
```

File: src/ctrl/nlas/mcast.rs (hunk 1 above)
Comment (dylanswiggett) [reply src/ctrl/nlas/mcast.rs]: Done, plus same for Op

File: src/ctrl/nlas/mcast.rs (hunk 2 above)
Comment (dylanswiggett) [reply src/ctrl/nlas/mcast.rs]: Ah, thanks for the reference. Updated based on your example.

File: src/ctrl/nlas/ops.rs (hunk 3 above)
Comment (dylanswiggett) [reply src/ctrl/nlas/ops.rs]: Fixed.

**Comment by dylanswiggett** [comment]
I've addressed your comments and updated the test cases to include proper indices and padding based on strace output.

File: src/ctrl/nlas/mod.rs [hunk 4]
```diff
@@ -140,8 +132,7 @@ impl<'a, T: AsRef<[u8]> + ?Sized> Parseable<NlaBuffer<&'a T>>
                         })
                     })
                     .collect::<Result<Vec<Vec<_>>, _>>()
-                    .context("failed to parse CTRL_ATTR_OPS")?;
-
+                    .context("failed to parse CTRL_ATTR_MCAST_GROUPS")?;
```
Comment (cathay4t) [inline src/ctrl/nlas/mod.rs:135]: Should be `CTRL_ATTR_OPS`

**Review by cathay4t — APPROVED** [review]
Looks good to me. I will fix the lint by append a patch.

File: src/ctrl/nlas/mod.rs (hunk 4 above)
Comment (cathay4t) [reply src/ctrl/nlas/mod.rs:135]: fixed in https://github.com/rust-netlink/netlink-packet-generic/pull/11

**Pull request merged**